db/hints/internal/hint_storage.cc

```diff
@@ -74,16 +73,14 @@ future<> scan_shard_hint_directories(const fs::path& hint_directory,
 ///                     K = <number of shards during the previous boot> * <number of endpoints
 ///                             for which hints where ever created>
 ///
-/// \note Should be called from a seastar::thread context.
-///
 /// \param hint_directory directory to scan
 /// \return a map: ep -> map: shard -> segments (full paths)
-hints_segments_map get_current_hints_segments(const fs::path& hint_directory) {
+future<hints_segments_map> get_current_hints_segments(const fs::path& hint_directory) {
     hints_segments_map current_hint_segments{};
 
     // Shard level.
-    scan_shard_hint_directories(hint_directory,
-            [&current_hint_segments] (fs::path dir, directory_entry de, unsigned shard_id) {
+    co_await scan_shard_hint_directories(hint_directory,
+            coroutine::lambda([&current_hint_segments] (fs::path dir, directory_entry de, unsigned shard_id) {
```

Comment:
You are using `coroutine::lambda` here as an argument to `scan_shard_hint_directories`, but not in `remove_irrelevant_shards_directories`. I don't think it is needed here as `lister::scan_dir` used in `scan_shard_hint_directories` will keep the std::function passed to it alive until it finishes its work.

Reply:
> You are using `coroutine::lambda` here as an argument to `scan_shard_hint_directories`, but not in `remove_irrelevant_shards_directories`.

That's because the lambda used in the latter function doesn't have any captures, so there's no need to wrap it in a `seastar::coroutine::lambda`.

> I don't think it is needed here as `lister::scan_dir` used in `scan_shard_hint_directories` will keep the `std::function` passed to it alive until it finishes its work.

But this is a very good point. I'll change it.